[ClangImporter] Check for builtin conformance in `importNumericLiteral`

hamishknight opened the pull request:
Make sure the destination type actually conforms to the corresponding builtin literal protocol before attempting to import it. We may want to consider allowing any type that conforms to the non-builtin literal protocol, but I want to keep this patch low risk and just ensure we at least don't crash for now.

rdar://156524292

hamishknight commented:
@swift-ci please test

hamishknight commented on test/ClangImporter/rdar156524292.swift:
```diff
@@ -0,0 +1,25 @@
+// RUN: %target-swift-frontend -typecheck -verify %t/main.swift -I %t -verify-additional-file %t/cmodule.h
+
+//--- cmodule.h
+#import <CoreGraphics/CoreGraphics.h>
```
Oops forgot to add a `REQUIRES` for this

slavapestov commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -125,6 +125,11 @@ static ValueDecl *importNumericLiteral(ClangImporter::Implementation &Impl,
 
+    auto &ctx = DC->getASTContext();
+    auto *constantTyNominal = constantType->getAnyNominal();
+    if (!constantTyNominal)
```
Should we add an assert at the other call sites too then, to avoid propagating nulls?

hamishknight commented on lib/ClangImporter/ImportMacro.cpp:
```diff
@@ -140,6 +145,16 @@ static ValueDecl *importNumericLiteral(ClangImporter::Implementation &Impl,
+      if (!ctx.getIntBuiltinInitDecl(constantTyNominal))
+        return nullptr;
+
       return createMacroConstant(Impl, MI, name, DC, constantType,
```
We cache the result on success, so it shouldn't matter much. There are also a bunch of different clients of `createConstant`, which would make it a bit awkward (although looking at them it's not entirely clear to me that we have the correct checks for them either)

hamishknight committed:
[AST] Also cache protocol kind in `getBuiltinInitDecl`

Make sure we don't return the wrong witness if you query with 2
different builtin protocol kinds.

hamishknight committed:
[AST] Remove asserts from `getBuiltinInitDecl`

Allow querying for types that may not conform. Failures here should
be asserted downstream (e.g in `setBuiltinInitializer`).

hamishknight committed:
[ClangImporter] Check for builtin conformance in `importNumericLiteral`

Make sure the destination type actually conforms to the corresponding
builtin literal protocol before attempting to import it. We may want
to consider allowing any type that conforms to the non-builtin literal
protocol, but I want to keep this patch low risk and just ensure we at
least don't crash for now.

rdar://156524292

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows
